File: samples/java/submit_bitcoin/SubmitBitcoin.java
/*
 * SecureFusion v1.0 — Bitcoin / OpenTimestamps submission example (Java).
 *
 * Submits a SecureFusion bundleHash to OpenTimestamps calendar servers,
 * which aggregate digests into Merkle trees and anchor them to Bitcoin.
 *
 * Uses java.net.http.HttpClient (JDK 11+) only — no third-party deps.
 *
 * Compile (with the SecureFusionAnchor sample on the classpath):
 *     cd samples/java
 *     javac SecureFusionAnchor.java submit_bitcoin/SubmitBitcoin.java
 *     java -cp .:submit_bitcoin SubmitBitcoin
 */

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.nio.file.*;
import java.time.Duration;
import java.util.*;

public class SubmitBitcoin {

    private static final String[] CALENDAR_SERVERS = {
        "https://a.pool.opentimestamps.org",
        "https://b.pool.opentimestamps.org",
        "https://finney.calendar.eternitywall.com",
    };

    private static final HttpClient HTTP = HttpClient.newBuilder()
        .connectTimeout(Duration.ofSeconds(15))
        .build();

    public static void main(String[] args) throws Exception {
        // 1. Load the example manifest and compute its bundleHash.
        Path examplesDir = findExamplesDir();
        String json = Files.readString(examplesDir.resolve("single-channel-event.json"), StandardCharsets.UTF_8);
        @SuppressWarnings("unchecked")
        Map<String, Object> manifest = (Map<String, Object>) SecureFusionAnchor.TinyJson.parse(json);
        SecureFusionAnchor.AnchorPayload payload = SecureFusionAnchor.buildAnchorPayload(manifest, null);

        System.out.println("SecureFusion v1.0 — OpenTimestamps (Bitcoin) submission");
        System.out.println("=".repeat(60));
        System.out.println("  bundleHash:  " + payload.bundleHash());
        System.out.println();

        // 2. Submit the digest to multiple calendars.
        byte[] digest = hexToBytes(payload.bundleHash());
        Map<String, byte[]> proofs = new LinkedHashMap<>();

        System.out.println("Anchoring to OpenTimestamps calendar servers:");
        for (String server : CALENDAR_SERVERS) {
            System.out.print("  Submitting to " + server + "... ");
            try {
                byte[] proof = submitToCalendar(digest, server);
                proofs.put(server, proof);
                System.out.println("[OK] (" + proof.length + " bytes)");
            } catch (Exception e) {
                System.out.println("[FAIL] " + e.getMessage());
            }
        }

        if (proofs.isEmpty()) {
            System.err.println();
            System.err.println("[FAIL] All OpenTimestamps calendars failed.");
            System.exit(1);
        }

        // 3. Save partial proofs.
        Path output = Path.of(payload.bundleHash().substring(0, 16) + ".partial.ots");
        try (var fos = Files.newOutputStream(output)) {
            for (var e : proofs.entrySet()) {
                fos.write(("--- " + e.getKey() + " ---\n").getBytes(StandardCharsets.UTF_8));
                fos.write(e.getValue());
                fos.write('\n');
            }
        }

        System.out.println();
        System.out.println("[OK] Saved partial proof: " + output);
        System.out.println();
        System.out.println("Next steps:");
        System.out.println("  1. The proof is currently 'partial' - calendars have aggregated");
        System.out.println("     your digest but Bitcoin has not yet committed to it.");
        System.out.println("  2. Wait at least 1 hour, then call the calendars' upgrade API");
        System.out.println("     to get the full Bitcoin block commitment.");
        System.out.println("  3. Store the upgraded .ots proof in the SecureFusion ledger.");
    }

    static byte[] submitToCalendar(byte[] digest, String calendarUrl) throws Exception {
        if (digest.length != 32)
            throw new IllegalArgumentException("OpenTimestamps digest must be 32 bytes");

        URI uri = URI.create(calendarUrl.replaceAll("/$", "") + "/digest");
        HttpRequest request = HttpRequest.newBuilder(uri)
            .header("Content-Type", "application/x-www-form-urlencoded")
            .timeout(Duration.ofSeconds(30))
            .POST(HttpRequest.BodyPublishers.ofByteArray(digest))
            .build();

        HttpResponse<byte[]> response = HTTP.send(request, HttpResponse.BodyHandlers.ofByteArray());
        if (response.statusCode() != 200)
            throw new RuntimeException("HTTP " + response.statusCode());
        return response.body();
    }

    private static Path findExamplesDir() {
        Path p = Path.of("").toAbsolutePath();
        for (int i = 0; i < 10 && p != null; i++, p = p.getParent()) {
            Path c = p.resolve("examples");
            if (Files.isDirectory(c)) return c;
        }
        throw new RuntimeException("Examples directory not found");
    }

    private static byte[] hexToBytes(String h) {
        byte[] out = new byte[h.length() / 2];
        for (int i = 0; i < h.length(); i += 2)
            out[i / 2] = (byte) ((Character.digit(h.charAt(i), 16) << 4) | Character.digit(h.charAt(i + 1), 16));
        return out;
    }
}

File: samples/csharp/submit_bitcoin/SubmitBitcoin.cs
// SecureFusion v1.0 — Bitcoin / OpenTimestamps submission example (C# / .NET 8).
//
// Submits a SecureFusion bundleHash to OpenTimestamps calendar servers,
// which aggregate digests into Merkle trees and anchor them to Bitcoin.
//
// OpenTimestamps provides a free, decentralised long-term anchor with the
// strongest legal precedent of any blockchain. Per-event cost: £0.
// Calendar aggregation latency: ~1 hour until Bitcoin commits.
//
// This file uses HttpClient directly — no third-party OTS library is
// required. For production use with full proof upgrading, consider porting
// the python-opentimestamps client logic, or call out to the `ots` CLI.

using System.Text.Json;
using SecureFusion.Samples.CSharp;

namespace SecureFusion.Samples.CSharp.SubmitBitcoin;

public static class BitcoinSubmitProgram
{
    private static readonly string[] CalendarServers = new[]
    {
        "https://a.pool.opentimestamps.org",
        "https://b.pool.opentimestamps.org",
        "https://finney.calendar.eternitywall.com",
    };

    private static readonly HttpClient Http = new() { Timeout = TimeSpan.FromSeconds(30) };

    public static async Task<int> Main()
    {
        // 1. Load the example manifest and compute its bundleHash.
        var examplesDir = FindExamplesDir();
        var manifestJson = await File.ReadAllTextAsync(Path.Combine(examplesDir, "single-channel-event.json"));
        using var doc = JsonDocument.Parse(manifestJson);
        var payload = AnchorBuilder.BuildAnchorPayload(doc.RootElement, applicationSigningKey: null);

        Console.WriteLine("SecureFusion v1.0 — OpenTimestamps (Bitcoin) submission");
        Console.WriteLine(new string('=', 60));
        Console.WriteLine($"  bundleHash:  {payload.BundleHash}");
        Console.WriteLine();

        // 2. Submit the digest to multiple calendar servers in parallel.
        var digest = Convert.FromHexString(payload.BundleHash);
        Console.WriteLine("Anchoring to OpenTimestamps calendar servers:");

        var proofs = new Dictionary<string, byte[]>();
        foreach (var server in CalendarServers)
        {
            Console.Write($"  Submitting to {server}... ");
            try
            {
                var proof = await SubmitToCalendar(digest, server);
                proofs[server] = proof;
                Console.WriteLine($"✓ ({proof.Length} bytes)");
            }
            catch (Exception ex)
            {
                Console.WriteLine($"✗ {ex.Message}");
            }
        }

        if (proofs.Count == 0)
        {
            Console.Error.WriteLine();
            Console.Error.WriteLine("✗ All OpenTimestamps calendars failed.");
            return 1;
        }

        // 3. Save the partial proofs.
        var outputPath = $"{payload.BundleHash[..16]}.partial.ots";
        await using var fs = File.Create(outputPath);
        foreach (var (url, proof) in proofs)
        {
            await fs.WriteAsync(System.Text.Encoding.UTF8.GetBytes($"--- {url} ---\n"));
            await fs.WriteAsync(proof);
            fs.WriteByte((byte)'\n');
        }

        Console.WriteLine();
        Console.WriteLine($"✓ Saved partial proof: {outputPath}");
        Console.WriteLine();
        Console.WriteLine("Next steps:");
        Console.WriteLine("  1. The proof is currently 'partial' — calendars have aggregated");
        Console.WriteLine("     your digest but Bitcoin has not yet committed to it.");
        Console.WriteLine("  2. Wait at least 1 hour, then call the calendars' upgrade API");
        Console.WriteLine("     to get the full Bitcoin block commitment.");
        Console.WriteLine("  3. Store the upgraded .ots proof in the SecureFusion ledger,");
        Console.WriteLine("     associated with the event record.");

        return 0;
    }

    private static async Task<byte[]> SubmitToCalendar(byte[] digest, string calendarUrl)
    {
        if (digest.Length != 32)
            throw new ArgumentException("OpenTimestamps digest must be 32 bytes (SHA-256)");

        var url = calendarUrl.TrimEnd('/') + "/digest";
        using var request = new HttpRequestMessage(HttpMethod.Post, url)
        {
            Content = new ByteArrayContent(digest),
        };
        request.Content.Headers.ContentType =
            new System.Net.Http.Headers.MediaTypeHeaderValue("application/x-www-form-urlencoded");

        using var response = await Http.SendAsync(request);
        response.EnsureSuccessStatusCode();
        return await response.Content.ReadAsByteArrayAsync();
    }

    private static string FindExamplesDir()
    {
        var dir = AppContext.BaseDirectory;
        for (int i = 0; i < 10; i++)
        {
            var candidate = Path.Combine(dir, "examples");
            if (Directory.Exists(candidate)) return candidate;
            dir = Path.GetDirectoryName(dir) ?? dir;
        }
        throw new DirectoryNotFoundException("Could not locate the SecureFusion examples directory");
    }
}
